bricks/stm32/main.c

```diff
@@ -57,7 +57,13 @@ void pb_stm32_poll(void) {
 }
 
 // callback for when stop button is pressed in IDE or on hub
-static void user_program_stop_func(void) {
+void pbsys_main_stop_program(void) {
+
+    // Don't stop if application program should not be interrupted.
+    if (mp_interrupt_char != CHAR_CTRL_C) {
+        return;
+    }
```

Comment:
This works fine, but I'll try generalizing it so it keeps working even if we enable `micropython.kbd_intr(chr)`

Reply:
The only reason we have this, by the way, is to avoid raising more exceptions while printing the exception. (I.e. what the `sys_unprepare` tried to do before).

> The exception message is printed and it use MicroPython print functions which can always raise an exception.
> The SystemExit has already been raised at this point, but there can be another exception.

But looking at how `mp_obj_print_exception` is used throughout MicroPython, it seems that it can just be called exactly like we do without additional protection.

So we can probably just remove this check altogether.

Reply:
This can be simplified further by removing the recursive call in `run_user_program` (which has superseded the `goto` from before) altogether. This makes the code simpler too. This comes without a build size increase for Move Hub since it doesn't have the REPL anyway.

Reply:
So now, it's simply:

```c
    // For MicroPython, the builtin program is the REPL.
    if (program->run_builtin) {
        // Init Pybricks package and auto-import everything.
        pb_package_pybricks_init(true);

        // Start the REPL.
        run_repl();    
    } else {
        // Init Pybricks package without auto-import.
        pb_package_pybricks_init(false);

        // Execute the user script.
        run_user_program(mpy_size, mpy_data);
    }
```

And

```c
static void run_user_program(uint32_t len, uint8_t *buf) {

    nlr_buf_t nlr;
    if (nlr_push(&nlr) == 0) {
        // <  run mpy >
        nlr_pop();
    } else {
        // Clear any pending exceptions (and run any callbacks).
        mp_hal_set_interrupt_char(-1);
        mp_handle_pending(false);

        // Print which exception triggered this.
        mp_obj_print_exception(&mp_plat_print, (mp_obj_t)nlr.ret_val);

        // If there was KeyboardInterrupt in the user program, drop to REPL
        // for debugging by running this function again with REPL enabled.
        if (mp_obj_exception_match((mp_obj_t)nlr.ret_val, &mp_type_KeyboardInterrupt)) {

            // The global scope is preserved to facilitate debugging, but we
            // soft-stop active resources like motors and sounds.
            pbio_stop_all(false);

            // Enter REPL.
            run_repl();
        }
    }
}
```